File: backend/src/Modules/Commitments/Controllers/MonthlyProgressController.cs
// Copyright (c) Quinntyne Brown. All Rights Reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using Commitments.Features.GoalProgress;
using Commitments.Features.MonthlyProgress;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Net;

namespace Commitments.Controllers;

[Authorize]
[ApiController]
[ApiVersion("1.0")]
[Route("api/v{version:apiVersion}/monthly-progress")]
[Produces(MediaTypeNames.Application.Json)]
[Consumes(MediaTypeNames.Application.Json)]
public class MonthlyProgressController
{
    private readonly ISender _sender;
    private readonly IHttpContextAccessor _httpContextAccessor;

    public MonthlyProgressController(IHttpContextAccessor httpContextAccessor, ISender sender)
    {
        _sender = sender;
        _httpContextAccessor = httpContextAccessor;
    }

    [HttpGet]
    [ProducesResponseType((int)HttpStatusCode.InternalServerError)]
    [ProducesResponseType(typeof(ProblemDetails), (int)HttpStatusCode.BadRequest)]
    [ProducesResponseType(typeof(GetMonthlyProgressResponse), (int)HttpStatusCode.OK)]
    public async Task<ActionResult<GetMonthlyProgressResponse>> Get([FromQuery] DateTimeOffset? asOf)
    {
        var response = _httpContextAccessor.HttpContext?.Response;
        if (response is not null)
            response.Headers.CacheControl = CacheControlPolicy.For(asOf, DateTimeOffset.UtcNow);

        return await _sender.Send(new GetMonthlyProgressRequest
        {
            ProfileId = _httpContextAccessor.GetProfileId(),
            AsOf = asOf
        });
    }
}

File: frontend/projects/commitments-dashboard-plugin/src/lib/tiles/monthly-progress-tile/monthly-progress.controller.ts
// Copyright (c) Quinntyne Brown. All Rights Reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { Injectable, computed, signal } from '@angular/core';
import { DashboardMode } from '@commitments/dashboard-framework';

import { MonthlyProgressDto, MonthlyProgressService } from '../../data/monthly-progress.service';

@Injectable()
export class MonthlyProgressController {
  readonly snapshot = signal<MonthlyProgressDto | null>(null);
  readonly mode = signal<DashboardMode>('live');

  readonly buckets = computed(() => this.snapshot()?.buckets ?? []);
  readonly isEmpty = computed(() => this.snapshot()?.isEmpty ?? true);

  constructor(private readonly _service: MonthlyProgressService) {}

  async load(mode: DashboardMode, asOf: string | null): Promise<void> {
    this.mode.set(mode);
    this.snapshot.set(await this._service.get(asOf));
  }
}
